data/transactions/verify/txn.go

```diff
@@ -121,13 +117,34 @@ func TxnGroup(stxs []transactions.SignedTxn, contextHdr bookkeeping.BlockHeader,
 	if err != nil {
 		return nil, err
 	}
+
+	minFeeCount := uint64(0)
+	feesPaid := uint64(0)
 	for i, stxn := range stxs {
 		err = Txn(&stxn, i, groupCtx)
 		if err != nil {
 			err = fmt.Errorf("transaction %+v invalid : %w", stxn, err)
 			return
 		}
+		if stxn.Txn.Type != protocol.CompactCertTx {
+			minFeeCount++
+		}
+		feesPaid = basics.AddSaturate(feesPaid, stxn.Txn.Fee.Raw)
 	}
+	feeNeeded, overflow := basics.OMul(groupCtx.consensusParams.MinTxnFee, minFeeCount)
+	if overflow {
+		err = fmt.Errorf("txgroup fee requirement overflow")
+		return
+	}
+	// feesPaid may have saturated. That's ok. Since we know
+	// feeNeeded did not overlfow, simple comparision tells us
+	// feesPaid was enough.
+	if feesPaid < feeNeeded {
+		err = fmt.Errorf("txgroup had %d in fees, which is less than the minimum %d * %d",
+			feesPaid, minFeeCount, groupCtx.consensusParams.MinTxnFee)
+		return
+	}
+
```

Comment:
The overflow fix looks good. But this entire code block needs to be under `proto.EnableFeePooling` switch, because in case of overflow a node with new code rejects the group and old non-upgraded node approves it since there is no such comparison. I understand this is not an issue on mainnet/testnet but still a possible problem on a some network with a supply larger than 2^64/10^6.

Reply:
I don't think that can happen.

It's not a matter of supply.  The only way this code overflows *and rejects* is if the total fee expected for the txgroup overflows on line 136 (not the fees paid).  That is, only if the network is expecting a fee for a single txgroup to be over 2^64.  I don't think such a txgroup can ever be paid for - rejection is inevitable.

The fees PAID can saturate without causing harm. So if the concern is someone deciding to screw around and put in a massive fee, far above the requirement, that won't cause a problem. It will saturate, and that will be higher than feeNeeded, so there's no rejection.  They'll end up paying the massive fee. (Though again, even this saturation is also impossible - there can't be more than 2^64 of supply, so how could so much be put into fees?)

If a txgroup's expected fee is > 2^64, that would be a problem, but no algorand network can have a supply high enough to pay such a fee. So if you configured a network to demand it, that txngroup is going to get rejected one way or another.

So I claim this check will never reject anything that the !EnableFeePooling accepts.

Reply:
you are right (again!). `feesPaid` can't overflow and it would be 2^64 in this case. Then rejection in `feesPaid < feeNeeded` will never happen.